### ISSUE TITLE
Optional Asterisk?

### ISSUE THREAD

**Issue by CWSpear** [AngularAgility/AngularAgility#97]
Is it possible to make an optional asterisk?

I tried using this strategy:

``` js
aaFormExtensionsProvider.labelStrategies.custom = function (element, labelText, isRequired) {
    var toggleRequired = element.attr('ng-required') 
        ? ' ng-show="' + element.attr('ng-required') + '"'
        : '';

    var label = angular.element('<label>')
        .attr('for', element[0].id)
        .addClass('control-label')
        .html(labelText + ' <span' + toggleRequired + '>*</span>');

    element.parent().parent().prepend(label);
};
```

which _would_ produce markup as expected:

``` html
<div class="field-group">
    <label id="{ID}" class="control-label">Some Field <span ng-show="needSomeField">*<span></label>
    <input ng-required="needSomeField" ...>
</div>
```

But that `ng-show` in the `label > span` won't "work" because the markup hasn't been `$compiled`. Is there a way to do this with the current code? If not, 2 solutions come to mind:

Either 1) pass the `$scope` to the custom strategy methods as another parameter so we can use our own `$compile` (gets tricky, because I don't think you can access `$compile` in `config`). Or, even with _just_ the `$scope`, we could set up our own `$watch`er to show/hide that span.

Or 2) run `$compile` on the element after it's built via the strategy definition method. You could run into backward compatibility issues with this, though, and infinite compile loops could be an issue.

**Comment by domderen** [comment]
Hey, you can actually get access to the scope of the element you are running on by doing 'element.scope()' in your label strategy. This is not the cleanest way to do it probably, but than again I would be against passing scope and injector into the strategies, and compiling code there. Instead I would work on the 'isRequired' parameter that is passed to the strategy, to be dynamic, and take into account if the required flag is conditional.

**Comment by johnculviner** [comment]
Good call @domderen . I didn't even think about that in this context. DI ruins your brain and makes you not think about statics haha :wink: Yeah it wouldn't be hard to just pass in $scope too and allow people to use templating rather than have to do stuff manually w/ direct DOM manipulation

**Comment by andrew-compeat** [comment]
I just solved this problem a little differently, and it seems to work. Angular Agility takes the returned labelElement and passes it to $compile for me, so I don't need $compile inside the strategy itself in order for the ng-class to work. (The ng-class did _not_ work until I added the return statement.)

CSS:

    .required:after {
        content: " *";
        color: red;
    }

JS:

    function myLabelStrategy(elements, labelText, isRequired) {
        var element, labelElement;

        if (elements && elements.length) {
            element = elements[0];

            if (element.type !== "hidden") {

                labelElement = $('<label>')
                    .attr({
                        "for": element.id,
                        "ng-class": "{ required: " + ($(element).attr("ng-required") || isRequired) + " }"
                    })
                    .html(labelText)[0];

                // For an input that's inside an input-group, put the label above the input-group rather than above the input.
                if ($(element.parentElement).hasClass("input-group")) {
                    element = element.parentElement;
                }

                element.parentNode.insertBefore(labelElement, element);

                return labelElement;
            }
        }
    };

**Comment by johnculviner** [comment]
I actually really like the class based approach. Can't believe I didn't think of it given how font awesome works the same way. I think if anything I might add another provider that just adds some additional classes based on the validation state to the group/label and give people easier CSS rather than programatic based control. Thanks!

**Comment by CWSpear** [comment]
Yeah... I too have actually *used* that approach before and really liked it, but for some reason totally forgot about it... but I saw @andrew-compeat's comment and even refactored some (non-Angular-Agility-related) code to use it today. =)

**Comment by johnculviner** [comment]
Nice. I think I'm going to remove the asterisk from the built in strategy even. No sense in not just using CSS to pull it off IMO
